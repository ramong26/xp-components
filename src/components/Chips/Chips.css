.chip {
  width: -moz-fit-content;
  width: fit-content;
  height: -moz-fit-content;
  height: fit-content;
  font-family: "Times New Roman", serif;
  padding: 4px 8px;
  border: 2px solid #000;
  border-radius: 0;
  background: #fffbea;
  background-size: cover;
  cursor: pointer;
  transition: all 0.15s ease-in-out;
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
  font-size: 13px;
  line-height: 1.69;
  font-weight: 800;
  color: #212121;
}
.chip:hover {
  background-size: cover;
  border-color: #c00;
  color: #c00;
  text-shadow: 1px 1px #fff;
}
.chip.selected {
  background: #ff0;
  border-color: #000;
  font-size: 13px;
  line-height: 1.69;
  font-weight: 900;
  color: #212121;
  text-transform: uppercase;
}
@media (min-width: 640px) {
  .chip {
    padding: 6px 12px;
    font-size: 14px;
    line-height: 1.71;
    font-weight: 800;
    color: #212121;
  }
  .chip.selected {
    font-size: 14px;
    line-height: 1.71;
    font-weight: 900;
    color: #212121;
  }
}
@media (min-width: 768px) {
  .chip {
    padding: 8px 16px;
    font-size: 14px;
    line-height: 1.71;
    font-weight: 900;
    color: #212121;
  }
}
@media (min-width: 1024px) {
  .chip {
    padding: 10px 18px;
    font-size: 16px;
    line-height: 1.625;
    font-weight: 900;
    color: #212121;
  }
}
@media (min-width: 1280px) {
  .chip {
    padding: 12px 20px;
    font-size: 16px;
    line-height: 1.625;
    font-weight: 900;
    color: #212121;
  }
}

.chip-label {
  margin-right: 4px;
}
@media (min-width: 640px) {
  .chip-label {
    margin-right: 6px;
  }
}
@media (min-width: 768px) {
  .chip-label {
    margin-right: 8px;
  }
}
@media (min-width: 1024px) {
  .chip-label {
    margin-right: 10px;
  }
}
@media (min-width: 1280px) {
  .chip-label {
    margin-right: 12px;
  }
}

.chip-remove {
  width: 14px;
  height: 14px;
  border: 2px solid #000;
  background: transparent;
  cursor: pointer;
  padding: 0;
  transition: all 0.15s ease-in-out;
  display: flex;
  flex-direction: row;
  justify-content: center;
  align-items: center;
  font-size: 13px;
  line-height: 1.69;
  font-weight: 900;
  color: #212121;
}
.chip-remove:hover {
  color: #c00;
  background: #fff;
}
@media (min-width: 640px) {
  .chip-remove {
    width: 16px;
    height: 16px;
    font-size: 14px;
    line-height: 1.71;
    font-weight: 900;
    color: #212121;
  }
}
@media (min-width: 768px) {
  .chip-remove {
    width: 18px;
    height: 18px;
    font-size: 14px;
    line-height: 1.71;
    font-weight: 900;
    color: #212121;
  }
}
@media (min-width: 1024px) {
  .chip-remove {
    width: 20px;
    height: 20px;
    font-size: 16px;
    line-height: 1.625;
    font-weight: 900;
    color: #212121;
  }
}
@media (min-width: 1280px) {
  .chip-remove {
    width: 22px;
    height: 22px;
    font-size: 16px;
    line-height: 1.625;
    font-weight: 900;
    color: #212121;
  }
}/*# sourceMappingURL=Chips.css.map */