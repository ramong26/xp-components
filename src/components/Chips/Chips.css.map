{"version":3,"sources":["Chips.scss","../../styles/_mixins.scss","Chips.css"],"names":[],"mappings":"AAGA;EACE,uBAAA;EAAA,kBAAA;EACA,wBAAA;EAAA,mBAAA;EACA,qCAAA;EACA,gBAAA;EACA,sBAAA;EACA,gBAAA;EACA,mBAAA;EACA,sBAAA;EACA,eAAA;EACA,iCAAA;ECqBA,aAAA;EACA,mBDrB2C;ECsB3C,8BDtBoB;ECuBpB,mBDvBmC;ECKnC,eAAA;EACA,iBAAA;EACA,gBDNmC;ECajC,cAAA;ACTJ;AFFE;EACE,sBAAA;EACA,kBAAA;EACA,WAAA;EACA,yBAAA;AEIJ;AFDE;EACE,gBAAA;EACA,kBAAA;ECPF,eAAA;EACA,iBAAA;EACA,gBDMqC;ECCnC,cAAA;EDAA,yBAAA;AEMJ;AD3BI;EDJJ;IA8BI,iBAAA;ICdF,eAAA;IACA,iBAAA;IACA,gBDaqC;ICNnC,cAAA;ECcF;EFNE;ICjBF,eAAA;IACA,iBAAA;IACA,gBDgBuC;ICTrC,cAAA;ECoBF;AACF;AD1CI;EDJJ;IAwCI,iBAAA;ICxBF,eAAA;IACA,iBAAA;IACA,gBDuBqC;IChBnC,cAAA;EC6BF;AACF;ADnDI;EDJJ;IA8CI,kBAAA;IC9BF,eAAA;IACA,kBAAA;IACA,gBD6BqC;ICtBnC,cAAA;ECsCF;AACF;AD5DI;EDJJ;IAoDI,kBAAA;ICpCF,eAAA;IACA,kBAAA;IACA,gBDmCqC;IC5BnC,cAAA;EC+CF;AACF;;AFhBA;EACE,iBAAA;AEmBF;ADzEI;EDqDJ;IAII,iBAAA;EEoBF;AACF;AD9EI;EDqDJ;IAOI,iBAAA;EEsBF;AACF;ADnFI;EDqDJ;IAUI,kBAAA;EEwBF;AACF;ADxFI;EDqDJ;IAaI,kBAAA;EE0BF;AACF;;AFvBA;EACE,WAAA;EACA,YAAA;EACA,sBAAA;EACA,uBAAA;EACA,eAAA;EACA,UAAA;EACA,iCAAA;EClDA,aAAA;EACA,mBDkDoC;ECjDpC,uBDiDoB;EChDpB,mBDgD4B;EClE5B,eAAA;EACA,iBAAA;EACA,gBDiEmC;EC1DjC,cAAA;AC0FJ;AF9BE;EACE,WAAA;EACA,gBAAA;AEgCJ;ADnHI;EDsEJ;IAkBI,WAAA;IACA,YAAA;IC7EF,eAAA;IACA,iBAAA;IACA,gBD4EqC;ICrEnC,cAAA;ECuGF;AACF;AD7HI;EDsEJ;IAyBI,WAAA;IACA,YAAA;ICpFF,eAAA;IACA,iBAAA;IACA,gBDmFqC;IC5EnC,cAAA;ECiHF;AACF;ADvII;EDsEJ;IAgCI,WAAA;IACA,YAAA;IC3FF,eAAA;IACA,kBAAA;IACA,gBD0FqC;ICnFnC,cAAA;EC2HF;AACF;ADjJI;EDsEJ;IAuCI,WAAA;IACA,YAAA;IClGF,eAAA;IACA,kBAAA;IACA,gBDiGqC;IC1FnC,cAAA;ECqIF;AACF","file":"Chips.css"}