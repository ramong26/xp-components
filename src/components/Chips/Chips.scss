@use '../../styles/index' as s;
@use 'sass:map';

$paper-texture: url('../../assets/paper.png');

.chip {
  width: fit-content;
  height: fit-content;
  font-family: 'Times New Roman', serif;
  padding: 0.3rem 0.8rem;
  border: 2px solid s.color(black);
  border-radius: 0;
  background: s.color(paper-background) $paper-texture;
  background-size: cover;
  cursor: pointer;
  transition: all 0.15s ease-in-out;
  @include s.flex-box(space-between, center, row);
  @include s.text-style-extended(md, 800, gray-900);

  &:hover {
    background-size: cover;
    border-color: s.color(red-0);
    color: s.color(red-0);
    text-shadow: 1px 1px s.color(white);
  }

  &.selected {
    background: s.color(yellow-0);
    border-color: s.color(black);
    @include s.text-style-extended(md, 900, gray-900);
    text-transform: uppercase;
  }
}

.chip-label {
  margin-right: 0.5rem;
}

.chip-remove {
  width: 1rem;
  height: 1rem;
  border: 2px solid s.color(black);
  background: transparent;
  cursor: pointer;
  padding: 0;
  transition: all 0.15s ease-in-out;
  @include s.flex-box(center, center, row);
  @include s.text-style-extended(md, 900, gray-900);
  &:hover {
    color: s.color(red-0);
    background: s.color(white);
  }
}
