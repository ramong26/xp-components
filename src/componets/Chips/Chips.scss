@use '../../styles/index' as s;
@use 'sass:map';

$paper-texture: url('../../assets/paper.png');

.chip {
  display: inline-flex;
  align-items: center;
  justify-content: space-between;
  padding: 0.3rem 0.8rem;
  border: 2px solid #000;
  border-radius: 0;
  background: #fffbea $paper-texture;
  background-size: cover;
  color: #111;
  font-size: 0.9rem;
  font-family: 'Times New Roman', serif;
  font-weight: bold;
  letter-spacing: 0.02em;
  cursor: pointer;
  transition: all 0.15s ease-in-out;

  &:hover {
    background: #fffbea $paper-texture;
    background-size: cover;
    border-color: #c00;
    color: #c00;
    text-shadow: 1px 1px #fff;
  }

  &.selected {
    background: #ff0;
    border-color: #000;
    color: #000;
    text-transform: uppercase;
    font-weight: 900;
    letter-spacing: 0.04em;
  }
}

.chip-label {
  margin-right: 0.5rem;
}

.chip-remove {
  width: 1rem;
  height: 1rem;
  display: flex;
  align-items: center;
  justify-content: center;
  color: #111;
  font-size: 0.75rem;
  font-weight: bold;
  border-left: 2px solid #000;
  background: transparent;
  cursor: pointer;
  padding: 0;

  &:hover {
    color: #c00;
    background: #fff;
  }
}
