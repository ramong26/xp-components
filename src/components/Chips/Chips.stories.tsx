import type { Meta, StoryObj } from '@storybook/react-vite';
import Chips from './Chips';
const meta: Meta<typeof Chips> = {
  title: 'Components/Chips',
  component: Chips,
  tags: ['autodocs'],
  argTypes: {
    onClick: { action: 'clicked' },
    onRemove: { action: 'removed' },
    selected: { control: 'boolean' }
  }
};

export default meta;

type Story = StoryObj<typeof Chips>;

// 기본 Chips 스토리
export const Default: Story = {
  args: {
    children: 'Chips Default',
    selected: false
  }
};

// 선택된 상태의 Chips 스토리
export const Selected: Story = {
  args: {
    children: 'Chips Selected',
    selected: true
  }
};

// 제거 기능이 있는 Chips 스토리
export const WithRemove: Story = {
  args: {
    children: 'Chips with Remove',
    onRemove: () => {},
    selected: false
  }
};
