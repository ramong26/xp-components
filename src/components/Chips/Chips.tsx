import React from 'react';
import './Chips.scss';

export interface ChipsProps extends React.HTMLAttributes<HTMLDivElement> {
  onRemove?: () => void;
  selected?: boolean;
  onClick?: () => void;
}

const Chips: React.FC<ChipsProps> = ({
  children,
  onRemove,
  selected,
  onClick,
  ...rest
}) => {
  return (
    <div
      className={`chip${selected ? ' selected' : ''}`}
      role={onClick ? 'button' : undefined}
      tabIndex={onClick ? 0 : undefined}
      onClick={onClick}
      onKeyDown={(e) => onClick && e.key === 'Enter' && onClick()}
      {...rest}
    >
      <span className="chip-label">{children}</span>
      {onRemove && (
        <button
          type="button"
          className="chip-remove"
          aria-label="Remove chip"
          onClick={(e) => {
            e.stopPropagation();
            onRemove();
          }}
        >
          Ã—
        </button>
      )}
    </div>
  );
};

export default Chips;
