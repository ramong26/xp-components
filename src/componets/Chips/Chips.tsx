import React from 'react';
import './Chips.scss';

export interface ChipsProps extends React.HTMLAttributes<HTMLDivElement> {
  onRemove?: () => void;
  selected?: boolean;
}

const Chips: React.FC<ChipsProps> = ({
  children,
  onRemove,
  selected,
  ...rest
}) => {
  return (
    <div className={`chip ${selected ? 'selected' : ''}`} {...rest}>
      <span className="chip-label">{children}</span>
      <button
        type="button"
        className="chip-remove"
        onClick={(e) => {
          e.stopPropagation();
          onRemove?.();
        }}
      >
        Ã—
      </button>
    </div>
  );
};

export default Chips;
