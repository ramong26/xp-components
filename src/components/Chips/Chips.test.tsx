vi.mock('../../assets/paper.png', () => 'paper.png');

import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { test, expect, vi, describe } from 'vitest';

import Chips from './Chips';

describe('Chips 컴포넌트', () => {
  test('Chips 렌더링 확인', () => {
    render(<Chips>Chips</Chips>);
    expect(screen.getByText('Chips')).toBeInTheDocument();
  });

  test('Chips 클릭 이벤트', async () => {
    const onClick = vi.fn();
    render(<Chips onClick={onClick}>Clickable Chips</Chips>);
    await userEvent.click(screen.getByText('Clickable Chips'));
    expect(onClick).toHaveBeenCalledTimes(1);
  });

  test('Chips 제거 버튼 클릭 이벤트', async () => {
    const onRemove = vi.fn();
    render(<Chips onRemove={onRemove}>Removable Chips</Chips>);
    const removeButton = screen.getByRole('button', { name: 'Remove chip' });
    await userEvent.click(removeButton);
    expect(onRemove).toHaveBeenCalledTimes(1);
  });

  test('selected prop에 따른 스타일 변경', () => {
    const { rerender, container } = render(<Chips>Default</Chips>);
    const chipElement = container.firstChild as HTMLElement;
    expect(chipElement).not.toHaveClass('selected');

    rerender(<Chips selected>Selected Chips</Chips>);
    const selectedChip = container.firstChild as HTMLElement;
    expect(selectedChip).toHaveClass('selected');
  });

  test('Chips 제거 버튼 클릭 시 onClick 이벤트 미발생', async () => {
    const onClick = vi.fn();
    const onRemove = vi.fn();
    render(
      <Chips onClick={onClick} onRemove={onRemove}>
        Clickable Chips
      </Chips>
    );
    const removeButton = screen.getByRole('button', { name: 'Remove chip' });
    await userEvent.click(removeButton);
    expect(onRemove).toHaveBeenCalledTimes(1);
    expect(onClick).not.toHaveBeenCalled();
  });

  test('onRemove가 없을 때 버튼 없음', () => {
    render(<Chips>Test Chips</Chips>);
    const removeButton = screen.queryByRole('button', { name: 'Remove chip' });
    expect(removeButton).toBeNull();
  });

  test('selected가 false일 때 클래스 없음', () => {
    const { container } = render(<Chips selected={false}>Not Selected</Chips>);
    const chipElement = container.firstChild as HTMLElement;
    expect(chipElement).not.toHaveClass('selected');
  });

  test('키보드 Enter로 클릭 이벤트 발생', async () => {
    const onClick = vi.fn();
    render(<Chips onClick={onClick}>Keyboard Chips</Chips>);
    const chipElement = screen.getByRole('button');
    chipElement.focus();
    await userEvent.keyboard('{Enter}');
    expect(onClick).toHaveBeenCalledTimes(1);
  });

  test('UI 스냅샷 테스트', () => {
    const { container } = render(<Chips selected>Snapshot Chips</Chips>);
    expect(container.firstChild).toMatchSnapshot();
  });
});
