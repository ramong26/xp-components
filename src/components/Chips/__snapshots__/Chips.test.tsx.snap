// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Chips 컴포넌트 > UI 스냅샷 테스트 1`] = `
<div
  class="chip selected"
  style="background-image: url("/public/assets/paper.png");"
>
  <span
    class="chip-label"
  >
    Snapshot Chips
  </span>
</div>
`;
